src/transform/src/table_elimination.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software matches_collection governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Detects an input being unioned with its negation and cancels them out
+
+use crate::{TransformArgs, TransformError};
+use expr::MirRelationExpr;
+
+/// Detects an input being unioned with its negation and cancels them out
+#[derive(Debug)]
+pub struct TableElimination;
+
+impl crate::Transform for TableElimination {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), TransformError> {
+        relation.try_visit_mut(&mut |e| self.action(e))
+    }
+}
+
+impl TableElimination {
+    /// Detects an input being unioned with its negation and cancels them out
+    pub fn action(&self, relation: &mut MirRelationExpr) -> Result<(), TransformError> {
+        if let MirRelationExpr::Union { base, inputs } = relation {
+            if let MirRelationExpr::Negate { input } = &**base {
+                if TableElimination::find_and_replace_relation(&input, inputs.iter_mut()) {
```

Comment:
One thing about rust is that you can only have one mutable reference to an object at a time, so you can't have:
```
for input in inputs.iter_mut() {
  if let MirRelationExpr::Negate {input: inner_input} = input {
     for input2 in inputs.iter_mut() {
        if input2 == inner_input {
          TableElimination::cancel_relation(input2);
        }
     }
  }
}
```
because `iter_mut()` takes a mutable reference on `inputs`, and you can't have two simultaneous mutable references on `inputs`.

We often get around this restriction by having iterating via index number instead of the objects themselves.
```
for i in 0..inputs.len() {
  let matching_negation = if let MirRelationExpr::Negate {input: inner_input} = inputs[i] {
     for j in 0..inputs.len() {
        if inputs[j] == inner_input {
          return Some(j)
        }
     } 
     None
  } else {
    None
  }
  if let Some(j) = matching_negation {
    TableElimination::cancel_relation(&mut inputs[i])
    TableElimination::cancel_relation(&mut inputs[j]);
  }
}
```
(You might able to find something simpler.)

Reply:
Oh, that's a lot simpler. Thank you!